plaso/parsers/postgresql.py

```diff
@@ -0,0 +1,182 @@
+  ).setResultsName('date_time')
+
+  _PID = pyparsing.Group(
+      pyparsing.Suppress('[') +
+      pyparsing.OneOrMore(text_parser.PyparsingConstants.INTEGER) +
+      pyparsing.Optional(pyparsing.Literal('-')) +
+      pyparsing.ZeroOrMore(text_parser.PyparsingConstants.INTEGER) +
+      pyparsing.Suppress(']')
+  ).setResultsName('pid')
+
+  _USER_AND_DATABASE = pyparsing.Group(
+      pyparsing.Word(pyparsing.alphanums) +
+      pyparsing.Literal('@') +
+      pyparsing.Word(pyparsing.alphanums)
+  ).setResultsName('user_and_database')
+
+  _LOG_LEVEL = (
+      pyparsing.Word(string.ascii_uppercase) +
+      pyparsing.Suppress(':')
+  ).setResultsName('log_level')
+
+  _LOG_LINE_END = _BODY_END = pyparsing.StringEnd() | _DATE_TIME
+
+  _LOG_LINE = (
+      _DATE_TIME +
+      _PID +
+      pyparsing.Optional(_USER_AND_DATABASE) +
+      _LOG_LEVEL +
+      pyparsing.SkipTo(_LOG_LINE_END).setResultsName('log_line') +
+      pyparsing.ZeroOrMore(pyparsing.lineEnd())
+  )
+
+  LINE_STRUCTURES = [('logline', _LOG_LINE)]
+
+  def _ConvertTimeString(self, structure):
+    """Converts the structure to a datetime object.
+
+    The date and time values are formatted as:
+    "2022-04-12 00:16:05.526 UTC".
+
+    Args:
+      structure (pyparsing.ParseResults): structure of tokens derived from a
+          line of a text file, that contains the time elements.
+
+    Returns:
+      datetime: datetime object.
+
+    Raises:
+      ValueError: if the structure cannot be converted into a datetime.
+    """
+    try:
+      dts = '{0:d}-{1:d}-{2:d} {3:d}:{4:d}:{5:d}.{6:d} {7:s}'.format(
+          structure['year'], structure['month'], structure['day_of_month'],
+          structure['time']['hours'], structure['time']['minutes'],
+          structure['time']['seconds'], structure.get('microseconds', [0])[0],
+          structure['time_zone'])
+    except (TypeError, ValueError) as exception:
+      raise ValueError(
+          'unable to format date time string with error: {0!s}.'.format(
+              exception))
+    return parser.parse(timestr=dts)
+
+
+  def ParseRecord(self, parser_mediator, key, structure):
+    """Parses a record and produces a PostgreSQL event.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfvfs.
+      key (str): name of the parsed structure.
+      structure (pyparsing.ParseResults): elements parsed from the file.
+
+    Raises:
+      ParseError: when the structure type is unknown.
+    """
+    if key != 'logline':
+      raise errors.ParseError(
+          'Unable to parse record, unknown structure: {0:s}'.format(key))
+
+    event_data = PostgreSQLEventData()
+    event_data.pid = ''.join(
+        [str(x) for x in self._GetValueFromStructure(structure, 'pid')])
+
+    log_level = self._GetValueFromStructure(structure, 'log_level')
+    if log_level and len(log_level) == 1:
```

Comment:
why not check for the inverse and return then continue with `event_data.log_level = log_level[0]` ?

Reply:
Done